File: src/application/excel/Excel.java
package application.excel;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.List;

import org.apache.poi.xssf.usermodel.XSSFCell;
import org.apache.poi.xssf.usermodel.XSSFCellStyle;
import org.apache.poi.xssf.usermodel.XSSFFont;
import org.apache.poi.xssf.usermodel.XSSFRow;
import org.apache.poi.xssf.usermodel.XSSFSheet;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

import application.controllers.MainMenu;

public class Excel
{

    public static XSSFCellStyle setDefaultStyle(XSSFWorkbook workbook)
    {
	XSSFFont defaultFont = workbook.createFont();
	XSSFCellStyle defaultStyle = workbook.createCellStyle();
	defaultFont.setFontHeightInPoints((short) 10);
	defaultFont.setFontName("Times New Roman");
	defaultFont.setBold(false);
	defaultStyle.setFont(defaultFont);

	return defaultStyle;
    }

    public static XSSFCellStyle setBoldStyle(XSSFWorkbook workbook)
    {
	XSSFCellStyle boldStyle = workbook.createCellStyle();
	XSSFFont boldFont = workbook.createFont();
	boldFont.setFontHeightInPoints((short) 12);
	boldFont.setFontName("Times New Roman");
	boldFont.setBold(true);
	boldFont.setBoldweight(XSSFFont.BOLDWEIGHT_BOLD);
	boldStyle.setFont(boldFont);

	return boldStyle;
    }

    public static OutputStream createOutputStream(String fileName)
    {

	try
	{
	    File file = new File(MainMenu.getProgramDirectory() + "/ChangeLists");
	    if (!file.exists())
	    {
		file.mkdir();
	    }
	    file = new File(MainMenu.getProgramDirectory() + "/ChangeLists/" + fileName);
	    System.out.println("File is: " + file.getAbsolutePath());
	    file.createNewFile();

	    return new BufferedOutputStream(new FileOutputStream(file));
	}
	catch (IOException e)
	{
	    // TODO Auto-generated catch block
	    e.printStackTrace();
	    return null;
	}
    }

    public static String CompareValues(int a, int b)
    {
	if (a == b && a >= 0) return Integer.toString(a);

	String aString;
	String bString;

	if (a < 0)
	    aString = "";
	else
	    aString = Integer.toString(a);
	if (b < 0)
	    bString = "";
	else
	    bString = Integer.toString(b);

	if (a < 0 && b < 0) return "";

	return aString + "=>" + bString;
    }

    public static String CompareValues(String a, String b)
    {
	if (a.equals(b))
	    return a;
	else
	    return a + "=>" + b;
    }

    public static void writeDataToSheet(XSSFSheet sheet, List<RowData> rowData)
    {
	for (int rowNum = 0; rowNum < rowData.size(); rowNum++)
	{
	    String[] data = rowData.get(rowNum).getText();
	    XSSFCellStyle style = rowData.get(rowNum).getStyle();
	    XSSFRow row = sheet.createRow(rowNum);

	    for (int cellNum = 0; cellNum < data.length; cellNum++)
	    {
		XSSFCell cell = row.createCell(cellNum);
		cell.getRow().setHeight((short) 0);
		cell.setCellStyle(style);
		cell.setCellValue(data[cellNum]);
	    }
	}
	int columnsInSheet = calculateNumberOfColumnsInSheet(sheet);
	// System.out.println("number of columns is:
	// "+Integer.toString(columnsInSheet));
	for (int i = 0; i < columnsInSheet; i++)
	{
	    sheet.autoSizeColumn(i);
	}
    }

    public static int calculateNumberOfColumnsInSheet(XSSFSheet sheet)
    {
	int numberOfColumns = 0;
	for (int i = 0; i < sheet.getLastRowNum(); i++)
	{
	    int columnsInRow = sheet.getRow(i).getLastCellNum();
	    if (columnsInRow > numberOfColumns)
	    {
		numberOfColumns = columnsInRow;
	    }
	}

	return numberOfColumns;
    }

    public static String modifyCase(String s)
    {
	if (s.isEmpty()) { return ""; }
	StringBuilder sb = new StringBuilder(s.toLowerCase().trim());
	sb.setCharAt(0, Character.toUpperCase(s.charAt(0)));

	for (int i = 0; i < sb.length(); i++)
	{
	    if (Character.isWhitespace(sb.charAt(i)))
	    {
		sb.setCharAt(i + 1, Character.toUpperCase(sb.charAt(i + 1)));
	    }
	}

	return sb.toString();
    }

    public static String Pget(List<String> t, int i)
    {
	if (i < t.size())
	{
	    return t.get(i);
	}
	else
	{
	    return "";
	}
    }

    public static void makeChangeLists()
    {
	CharacterData.makeChangeList();
	EquipmentData.makeChangeList();
	CharacterComparisons.makeChangeList();
	ExpData.makeChangeList();
	MonsterData.makeChangeList();
	// ApprenticeData.makeChangeList();
	// DropTableData.makeChangeList();

	// ItemData.makeChangeList();

	// ShopData.makeChangeList();
	// SpellData.makeChangeList();

    }

}

File: src/application/excel/Excel.kt
package application.excel

import application.ROM
import org.apache.poi.xssf.usermodel.XSSFCellStyle
import org.apache.poi.xssf.usermodel.XSSFSheet
import org.apache.poi.xssf.usermodel.XSSFWorkbook
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.lang.Integer.toString

open class Excel(var fileName: String)
{
	val workbook = XSSFWorkbook()
	val defaultStyle: XSSFCellStyle = workbook.createCellStyle()
	val bigStyle: XSSFCellStyle = workbook.createCellStyle()
	val boldStyle: XSSFCellStyle = workbook.createCellStyle()
	val bigBoldStyle: XSSFCellStyle = workbook.createCellStyle()


	init
	{
		setDefaultStyle()
		setBigStyle()
		setBoldStyle()
		setBigBoldStyle()
		saveData()
	}

	private fun setDefaultStyle()
	{
		val defaultFont = workbook.createFont()
		defaultFont.fontHeightInPoints = 14.toShort()
		defaultFont.fontName = "Times New Roman"
		defaultFont.bold = false
		defaultStyle.setFont(defaultFont)
	}

	fun setBigStyle()
	{
		val boldFont = workbook.createFont()
		boldFont.fontHeightInPoints = 16.toShort()
		boldFont.fontName = "Times New Roman"
		boldFont.bold = false
		bigStyle.setFont(boldFont)
	}

	fun setBoldStyle()
	{
		val boldFont = workbook.createFont()
		boldFont.fontHeightInPoints = 14.toShort()
		boldFont.fontName = "Times New Roman"
		boldFont.bold = true
		boldStyle.setFont(boldFont)
	}
	fun setBigBoldStyle()
	{
		val boldFont = workbook.createFont()
		boldFont.fontHeightInPoints = 16.toShort()
		boldFont.fontName = "Times New Roman"
		boldFont.bold = true
		bigBoldStyle.setFont(boldFont)
	}

	fun CompareValues(a: Int, b: Int): String
	{
		if (a == b && a >= 0) return toString(a)

		val aString: String
		val bString: String

		if (a < 0) aString = ""
		else aString = toString(a)
		if (b < 0) bString = ""
		else bString = toString(b)

		if (a < 0 && b < 0) return ""

		return aString + "=>" + bString
	}

	fun CompareValues(a: String, b: String): String
	{
		if (a == b) return a
		else return a + "=>" + b
	}

	fun writeDataToSheet(sheet: XSSFSheet, sheetData: SheetData)
	{
		sheetData.rows.forEachIndexed {rowIndex, rowData ->
			val row = sheet.createRow(rowIndex)

			rowData.cells.forEachIndexed{ cellIndex, cellData ->
				val cell = row.createCell(cellIndex)
				cell.cellStyle = cellData.style
				cell.setCellValue(cellData.text)
			}
		}

		val columnsInSheet = calculateNumberOfColumnsInSheet(sheet)
		// System.out.println("number of columns is:
		// "+Integer.toString(columnsInSheet));
		for (i in 0 until columnsInSheet) sheet.autoSizeColumn(i)
	}

	fun calculateNumberOfColumnsInSheet(sheet: XSSFSheet): Int
	{
		var numberOfColumns = 0
		(0..sheet.lastRowNum - 1).forEach {
			val columnsInRow = sheet.getRow(it).lastCellNum.toInt()
			if (columnsInRow > numberOfColumns) numberOfColumns = columnsInRow
		}
		return numberOfColumns
	}

	fun modifyCase(s: String): String
	{
		if (s.isEmpty()) return ""
		val sb = StringBuilder(s.toLowerCase().trim { it <= ' ' })
		sb.setCharAt(0, Character.toUpperCase(s[0]))

		(0..sb.length - 1).forEach { if (Character.isWhitespace(sb[it])) sb.setCharAt(it + 1, Character.toUpperCase(sb[it + 1])) }

		return sb.toString()
	}

	fun Pget(t: List<String>, i: Int): String
	{
		if (i < t.size) return t[i]
		else return ""
	}

	fun saveData()
	{
		makeChangeList()
		try
		{
			var file = ROM.changeListsDirectory
			println("Changelist Directory: " + file)
			file.mkdirs()

			file = File(file.path + "/" + fileName)
			println("Changelist File Location: " + file.absolutePath)
			file.createNewFile()
			BufferedOutputStream(FileOutputStream(file)).use {
				workbook.write(it)
			}
		}
		catch (e: IOException)
		{
			// TODO Auto-generated catch block
			e.printStackTrace()
		}
	}

	open fun makeChangeList()
	{
	}

}
